evennia/objects/objects.py

```diff
@@ -1889,6 +1889,121 @@ def at_say(self, message, msg_self=None, msg_location=None,
+        message = list()
+
+        if not (len(self.cmdset.all()) == 1 and self.cmdset.current.key == "_EMPTY_CMDSET"):
+            # all() returns a 'stack', so make a copy to sort.
+            stored_cmdsets = sorted(self.cmdset.all(), key=lambda x: x.priority, reverse=True)
+            message.append("|wStored Cmdset(s)|n:\n %s" % ("\n ".join("%s [%s] (%s, prio %s)" % (
+                cmdset.path, cmdset.key, cmdset.mergetype, cmdset.priority) for cmdset in stored_cmdsets
+                                                                      if cmdset.key != "_EMPTY_CMDSET")))
+
+            # this gets all components of the currently merged set
+            all_cmdsets = [(cmdset.key, cmdset) for cmdset in cmdsets.merged_from]
+            # we always at least try to add account- and session sets since these are ignored
+            # if we merge on the object level.
+            if hasattr(self, "account") and self.account:
+                all_cmdsets.extend([(cmdset.key, cmdset) for cmdset in self.account.cmdset.all()])
+                if self.sessions.count():
+                    # if there are more sessions than one on objects it's because of multisession mode 3.
+                    # we only show the first session's cmdset here (it is -in principle- possible that
+                    # different sessions have different cmdsets but for admins who want such madness
+                    # it is better that they overload with their own CmdExamine to handle it).
+                    all_cmdsets.extend([(cmdset.key, cmdset) for cmdset in self.account.sessions.all()[0].cmdset.all()])
+            else:
+                try:
+                    # we have to protect this since many objects don't have sessions.
+                    all_cmdsets.extend([(cmdset.key, cmdset)
+                                        for cmdset in self.get_session(self.sessions.get()).cmdset.all()])
+                except (TypeError, AttributeError):
+                    # an error means we are merging an object without a session
+                    pass
+            all_cmdsets = [cmdset for cmdset in dict(all_cmdsets).values()]
+            all_cmdsets.sort(key=lambda x: x.priority, reverse=True)
+            message.append("|wMerged Cmdset(s)|n:\n %s" % ("\n ".join("%s [%s] (%s, prio %s)" % (
+                cmdset.path, cmdset.key, cmdset.mergetype, cmdset.priority) for cmdset in all_cmdsets)))
+
+            # list the commands available to this object
+            avail_cmdset = sorted([cmd.key for cmd in cmdsets
+                                   if cmd.access(self, "cmd")])
+            cmdsetstr = fill(", ".join(str(cmd) for cmd in cmdsets), indent=2)
+            message.append("|wCommands available to %s (result of Merged CmdSets)|n:\n %s" % (self.key, cmdsetstr))
+
+        return '\n'.join(str(l) for l in message)
+
+    def return_examine_contents(self, viewer, *args, **kwargs):
```

Comment:
This should also not be a new hook. Either embed it in `return_examine` or maybe put it on the ContentsHandler?